dask/dataframe/core.py

```diff
@@ -2330,6 +2335,257 @@ def quantile(df, q):
+        groups = tree_groups(prev_width, width)
+        keys = [(token, level, i) for i in range(width)]
+        rv.update((key, (func, list(take(num, prev_keys))))
+                   for num, key in zip(groups, keys))
+        prev_width = width
+        prev_keys = iter(keys)
+        level += 1
+    return rv
+
+
+def _merge_sorted(items):
+    return list(merge_sorted(*items))
+
+
+def _prepare_percentile_merge(qs, vals, length):
+    """Weigh percentile values by length and the difference between percentiles
+
+    >>> percentiles = [0, 25, 50, 90, 100]
+    >>> values = [2, 3, 5, 8, 13]
+    >>> length = 10
+    >>> _prepare_percentile_merge(percentiles, values, length)
+    [(2, 125.0), (3, 250.0), (5, 325.0), (8, 250.0), (13, 50.0)]
+
+    The weight of the first element, ``2``, is determined by the difference
+    between the first and second percentiles, and then scaled by length:
+
+    >>> 0.5 * length * (percentiles[1] - percentiles[0])
+    125.0
+
+    The second weight uses the difference of percentiles on both sides, so
+    it will be twice the first weight if the percentiles are equally spaced:
+
+    >>> 0.5 * length * (percentiles[2] - percentiles[0])
+    250.0
+
+    """
+    if length == 0:
+        return []
+    diff = np.ediff1d(qs, 0.0, 0.0)
+    weights = 0.5 * length * (diff[1:] + diff[:-1])
+    return list(zip(vals, weights))
+
+
+def _process_val_weights(vals_and_weights, finalq):
+    """Calculate final percentiles given weighted vals"""
+    vals, weights = zip(*vals_and_weights)
+    vals = np.array(vals)
+    weights = np.array(weights)
+
+    # percentile-like, but scaled by weights
+    q = np.cumsum(weights)
+
+    # rescale finalq percentiles to match q
+    desired_q = finalq * q[-1]
+
+    if np.issubdtype(vals.dtype, np.number):
+        rv = np.interp(desired_q, q, vals)
+    else:
+        left = np.searchsorted(q, desired_q, side='left')
+        right = np.searchsorted(q, desired_q, side='right') - 1
+        # stay inbounds
+        np.maximum(right, 0, right)
+        lower = np.minimum(left, right)
+        rv = vals[lower]
+    return rv
+
+
+def _weighted_percentiles(df, num_old, num_new, upsample=1.0, random_state=None):
+    from dask.array.percentile import _percentile
+    length = len(df)
+    qs = _sample_percentiles(num_old, num_new, length, upsample, random_state)
+    vals = _percentile(df.values, qs)
+    vals_and_weights = _prepare_percentile_merge(qs, vals, length)
+    return vals_and_weights
+
+
+def _repartition_quantiles(df, npartitions, upsample=1.0, random_state=None):
+    """ Approximate quantiles of Series used for repartitioning
+    """
+    assert isinstance(df, Series)
+    from dask.array.percentile import _percentile
+
+    qs = np.linspace(0, 1, npartitions + 1)
+    # currently, only Series has quantile method
+    # Index.quantile(list-like) must be pd.Series, not pd.Index
+    merge_type = lambda v: pd.Series(v, index=qs, name=df.name)
```

Comment:
This fails in a distributed setting because `df`, a dask.dataframe, is closed over inside of a method being sent over a wire